pkg/api/steve/collections/apiresources.go

```diff
@@ -0,0 +1,172 @@
+		client:      discovery,
+		mapper:      mapper,
+		resourceMap: make(map[string]metav1.APIResource),
+	}
+
+	crd.OnChange(ctx, "collections-api", a.OnChangeCRD)
+	apiService.OnChange(ctx, "collections-api", a.OnChangeAPIService)
+	return a
+}
+
+// OnChangeCRD queues the refresh when a CRD event occurs.
+func (a *apiResourceWatcher) OnChangeCRD(_ string, crd *apiextv1.CustomResourceDefinition) (*apiextv1.CustomResourceDefinition, error) {
+	a.queueRefresh()
+	return crd, nil
+}
+
+// OnChangeAPIService queues the refresh when an APIService event occurs.
+func (a *apiResourceWatcher) OnChangeAPIService(_ string, apiService *apiregv1.APIService) (*apiregv1.APIService, error) {
+	a.queueRefresh()
+	return apiService, nil
+}
+
+// GetKindForResource returns the resource Kind given its GVR.
+func (a *apiResourceWatcher) GetKindForResource(gvr schema.GroupVersionResource) (string, error) {
+	a.RLock()
+	defer a.RUnlock()
+	gvk, err := a.mapper.KindFor(gvr)
+	if err != nil {
+		return "", fmt.Errorf("failed to look up kind for resource %s, %w", gvr.Resource, err)
+	}
+	return gvk.Kind, nil
+}
+
+// List returns all the APIResources for the collections API.
+func (a *apiResourceWatcher) List() []metav1.APIResource {
+	a.RLock()
+	defer a.RUnlock()
+	return a.apiResources
+}
+
+// Get returns an APIResource and an existence bool given a resource and group.
+func (a *apiResourceWatcher) Get(resource, group string) (metav1.APIResource, bool) {
+	a.RLock()
+	defer a.RUnlock()
+	if group == "" {
+		val, ok := a.resourceMap[resource]
+		return val, ok
+	}
+	val, ok := a.resourceMap[group+"."+resource]
+	return val, ok
+}
+
+func (a *apiResourceWatcher) queueRefresh() {
+	a.toSync.Store(1)
+
+	go func() {
+		time.Sleep(queueRefreshDelay * time.Millisecond)
+		if err := a.refreshAll(); err != nil {
+			logrus.Errorf("[%s] failed to sync schemas, will retry: %v", apiServiceName, err)
+			a.toSync.Store(1)
+		}
+	}()
+}
+
+func (a *apiResourceWatcher) setAPIResources() error {
+	resourceList, err := discovery.ServerPreferredNamespacedResources(a.client)
+	if err != nil {
+		return err
+	}
+	result := []metav1.APIResource{}
+	a.Lock()
+	defer a.Unlock()
+	for _, resource := range resourceList {
+		if resource.GroupVersion == groupVersion {
+			continue
+		}
+		gv := strings.Split(resource.GroupVersion, "/")
+		group := ""
+		version := ""
+		if len(gv) > 1 {
+			group = gv[0]
+			version = gv[1]
+		} else {
+			version = gv[0]
+		}
+		for _, r := range resource.APIResources {
+			name := r.Name
+			if group != "" {
+				name = group + "." + name
+			}
+			resource := metav1.APIResource{
+				Name:               name,
+				Group:              group,
+				Version:            version,
+				Kind:               r.Kind,
+				Verbs:              []string{"list"},
+				Namespaced:         true,
+				StorageVersionHash: apidiscovery.StorageVersionHash(group, version, r.Kind),
+			}
+			result = append(result, resource)
+			a.resourceMap[name] = resource
```

Comment:
There's nothing that cleans it up. I'll add that. Any leftover resources will increase the size of the map but won't otherwise cause any issue, the authoritative list that the API server uses is `a.apiResources`, this map is used to get individual resources in `Get()`.

Reply:
Adding a reset for this.